metals/src/main/scala/scala/meta/internal/metals/PackageProvider.scala

```diff
@@ -169,139 +199,195 @@ class PackageProvider(
-            refsInImportsEdits ++ newImports
-        }
-        new WorkspaceEdit(
-          Map(filePath.toURI.toString -> changes.toList.asJava).asJava
-        )
+      oldPkgParts: List[String],
+      newPkgParts: List[String],
+      references: List[Reference],
+      fileFilter: String => Boolean,
+  ): WorkspaceEdit =
+    references
+      .groupBy(_.uri)
+      .withFilter { case (uri, _) => fileFilter(uri) }
+      .map { case (uri, references) =>
+        calcRefsEditsForFile(oldPkgParts, newPkgParts, references, uri)
       }
-    edits.toSeq.mergeChanges
+      .toSeq
+      .mergeChanges
+
+  private def calcRefsEditsForFile(
+      oldPkgParts: List[String],
+      newPkgParts: List[String],
+      references: List[Reference],
+      fileUri: String,
+  ): WorkspaceEdit = {
+    val filePath = fileUri.toAbsolutePath
+    val optTree = trees.get(filePath)
+    val referencesSymbols = references.flatMap(_.symbols).toSet
+
+    val importersAndParts = optTree
+      .fold(Vector.empty[Importer])(findImporters)
+      .map(i => (i, extractNames(i.ref)))
+
+    val baseRename = PkgPartsRename(oldPkgParts, newPkgParts)
+    val packagesRenames = optTree
+      .map(collectPkgRenames(baseRename, _))
+      .getOrElse(List(baseRename))
+    val shortestRename = packagesRenames.last.newPkgParts.mkString(".")
+
+    val referencesNames = ReferencesNames(references.toList)
+    // handles fully qualified references in selects:
+    // 1. import pkg1.pkg2.pkg3.Object.value
+    // 2. pkg1.pk2.pkg3.Object
+    // 3. pkg1.pk2.pkg3.Object.value
+    val ImportEdits(handledByFullyQualified, fullyQuilifiedChanges) =
+      calcFullyQuilifiedEdits(references.toList, filePath, packagesRenames)
+    // handles package renames in imports
+    // deletes wildcard import if it can be deleted
+    val ImportEdits(imported, refsInImportsEdits) =
+      calcAllImportsEditsForFileMove(
+        importersAndParts,
+        referencesNames,
+        referencesSymbols,
+        filePath,
+        packagesRenames,
+      )
+    val toImport = (references.toSet -- handledByFullyQualified)
+      .map(_.definition.name)
+      .toSet -- imported
+    // handles adding new imports
+    val newImports = Option.when(toImport.nonEmpty) {
+      calcNewImports(optTree, toImport, shortestRename)
+    }
+    val changes =
+      refsInImportsEdits ++ newImports ++ fullyQuilifiedChanges
+    new WorkspaceEdit(
+      Map(filePath.toURI.toString -> changes.toList.asJava).asJava
+    )
   }
 
-  private def calcImportEditsForDirMove(
-      importer: Importer,
-      importParts: Vector[String],
-      oldPkgPrefix: Vector[String],
-      newPkgPrefix: Vector[String],
-  ): List[TextEdit] = if (importParts.startsWith(oldPkgPrefix)) {
-    val newImport = newPkgPrefix ++ importParts.drop(oldPkgPrefix.size)
-    List(new TextEdit(importer.ref.pos.toLsp, newImport.mkString(".")))
-  } else List.empty
+  // This function collects possible pakages renames for a file.
```

Comment:
This seems a bit unclear, could you maybe add scala file snippets? Might help to understand it better.

Also a nitpick, could you convert it to a normal scaladoc comment?